docker/docker.ts

```diff
@@ -346,25 +346,48 @@ async function dockerBuild(
     await runCommandThatMustSucceed("docker", buildArgs, logResource);
 }
 
+// Keep track of registries and users that have been logged in.  If we've already logged into that
+// registry with that user, there's no need to do it again.
+const loggedInUsers: { registryName: string, username: string }[] = [];
```

Comment:
I deeply fear global state in Node modules - is there a way we can attach this to the registry?

Reply:
the registry, in this case, is really just a triple of server/username/password.  There's no real instance anywhere that we woudl know would be "the one" to use to store data on.  

We could consider changing the form of the 'docker' package to be more about exposing a richer object that itself could then cache this data.  cloud-aws and auzre would then create a singleton of that object that they would then use.  

However... this seems like a lot of work.  I'm curious what @lukehoban thinks about things here and if we should be rethinking how the docker layer potentially works here wrt to statefulness...